### PR TITLE
=htt #20208 empty responses cannot be consumed without size limit

### PR DESCRIPTION
**Pull request by hochgi**
This happens when entity responses are received as `HttpEntity.Default` with empty body.
an `IllegalArgumentException` is thrown with the message: `requirement failed: contentLength must be positive (use `HttpEntity.empty(contentType)` for empty entities)`

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Refs #20208, thanks for reporting

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 1]
```diff
@@ -447,6 +447,37 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
...
+
+        val serverOutSub = serverOut.expectSubscription()
+        serverOutSub.expectRequest()
+        serverOutSub.sendNext(HttpResponse(StatusCodes.OK, List(RawHeader("X-Res-Type", "EMPTY")), HttpEntity.Default(ContentTypes.`text/plain(UTF-8)`, 0L, Source.empty[ByteString])))
+
+        val clientInSub = clientIn.expectSubscription()
+        clientInSub.request(1)
+        val response = clientIn.expectNext()
+        val unlimitedEmptyFut = response.entity.withSizeLimit(-1).dataBytes.runFold(ByteString(""))(_ ++ _)
+        Await.result(unlimitedEmptyFut, 100.millis).utf8String shouldEqual ""
```
Comment (patriknw) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: That timeout is too short for some of our CI environments. If there is no reason for a short timeout it should be 3 seconds. Please use `remaining` or `remainingOrDefault`.

**Comment by ktoso** [comment]
Uhm, but this test doesn't make sense – the faulure here is explicit validation code that explains that one should not use HttpEntity.Default when the entity is known empty, exactly as the error states: contentLength must be positive **(use `HttpEntity.empty(contentType)` for empty entities)**

It never even reaches the consuming part.
Did you see this error outside of a test scenario? If so then we likely need to fix the origin where such empty Default entity is created instead.

**Comment by hochgi** [comment]
@ktoso yes, I actually figured out about the test not making much sense, but was swamped at work, and haven't found the time to fix it yet.
I did get the exact same exception at runtime using client API.
The server side is play 2.4.6, and the code that generated the response in play is basically reduced to:
```scala
Ok("").withHeaders("X-SOME-SPECIAL-HEADER" -> "something")
```

**Comment by ktoso** [comment]
Thanks for the info, I should be able to reproduce then!

**Comment by ktoso** [comment]
I am unable to reproduce, full details:

Play provided response, nothing weird here:

```
$ http 127.0.0.1:9000
HTTP/1.1 200 OK
Content-Length: 0
Content-Type: text/plain; charset=utf-8
Date: Tue, 05 Apr 2016 14:04:46 GMT
X-ExampleFilter: foo
X-SOME-SPECIAL-HEADER: something


```

full code to attempt reproducing:

```scala

import akka.Done
import akka.http.scaladsl.marshallers.xml.ScalaXmlSupport
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.directives.Credentials
import com.typesafe.config.{ Config, ConfigFactory }
import akka.actor.ActorSystem
import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
import akka.http.scaladsl.Http
import akka.stream.scaladsl.Sink
import akka.util.ByteString

import scala.concurrent.{ Await, Future }
import scala.concurrent.duration._

object ReproducerClientApp extends App {
  val testConf: Config = ConfigFactory.parseString("""
    akka.loglevel = DEBUG
    akka.log-dead-letters = off
    akka.stream.materializer.debug.fuzzing-mode = off
    """)
  implicit val system = ActorSystem("Reproducer", testConf)
  implicit val materializer = ActorMaterializer()
  import system.dispatcher

  val request =
    HttpRequest(uri = "http://127.0.0.1:9000")

  val response =
    Http().singleRequest(request).flatMap { rs ⇒
      println("rs.entity = " + rs.entity)
      println("rs = " + rs.entity)

      rs.entity.withSizeLimit(-1).dataBytes.runFold(ByteString(""))(_ ++ _)
    }

  import scala.concurrent.duration._
  println(Await.result(response, 10 seconds))
  Http().shutdownAllConnectionPools().flatMap(x ⇒ system.terminate())
}
```

Response is valid, we yield just the empty byte string:

```
[info] Running akka.http.scaladsl.server.ReproducerClientApp
[DEBUG] [04/05/2016 16:05:15.935] [run-main-c] [EventStream(akka://Reproducer)] logger log1-TestEventListener started
[DEBUG] [04/05/2016 16:05:15.938] [run-main-c] [EventStream(akka://Reproducer)] Default Loggers started
[DEBUG] [04/05/2016 16:05:16.309] [run-main-c] [AkkaSSLConfig(akka://Reproducer)] Initializing AkkaSSLConfig extension...
[DEBUG] [04/05/2016 16:05:16.348] [run-main-c] [AkkaSSLConfig(akka://Reproducer)] buildHostnameVerifier: created hostname verifier: com.typesafe.sslconfig.ssl.DefaultHostnameVerifier@62fbb4d5
[DEBUG] [04/05/2016 16:05:16.714] [Reproducer-akka.actor.default-dispatcher-3] [akka://Reproducer/user/PoolInterfaceActor-0] (Re-)starting host connection pool to 127.0.0.1:9000
[DEBUG] [04/05/2016 16:05:17.494] [Reproducer-akka.actor.default-dispatcher-4] [akka://Reproducer/user/SlotProcessor-1] become unconnected, from subscriber pending
[DEBUG] [04/05/2016 16:05:17.494] [Reproducer-akka.actor.default-dispatcher-7] [akka://Reproducer/user/SlotProcessor-2] become unconnected, from subscriber pending
[DEBUG] [04/05/2016 16:05:17.494] [Reproducer-akka.actor.default-dispatcher-8] [akka://Reproducer/user/SlotProcessor-3] become unconnected, from subscriber pending
[DEBUG] [04/05/2016 16:05:17.494] [Reproducer-akka.actor.default-dispatcher-5] [akka://Reproducer/user/SlotProcessor-0] become unconnected, from subscriber pending
[DEBUG] [04/05/2016 16:05:17.596] [Reproducer-akka.actor.default-dispatcher-4] [akka://Reproducer/system/IO-TCP/selectors/$a/0] Attempting connection to [/127.0.0.1:9000]
[DEBUG] [04/05/2016 16:05:17.604] [Reproducer-akka.actor.default-dispatcher-4] [akka://Reproducer/system/IO-TCP/selectors/$a/0] Connection established to [/127.0.0.1:9000]
rs.entity = HttpEntity.Strict(text/plain; charset=UTF-8,)
rs = HttpEntity.Strict(text/plain; charset=UTF-8,)
ByteString()
[DEBUG] [04/05/2016 16:05:17.759] [run-main-c] [akka.actor.ActorSystemImpl(Reproducer)] Initiating orderly shutdown of all active host connections pools...
[DEBUG] [04/05/2016 16:05:17.769] [Reproducer-akka.actor.default-dispatcher-4] [akka://Reproducer/user/PoolInterfaceActor-0] Shutting down host connection pool to 127.0.0.1:9000
[DEBUG] [04/05/2016 16:05:17.779] [Reproducer-akka.actor.default-dispatcher-2] [akka://Reproducer/user/SlotProcessor-0] Slot 0 disconnected after regular connection close
[DEBUG] [04/05/2016 16:05:17.783] [Reproducer-akka.actor.default-dispatcher-7] [akka://Reproducer/user/PoolInterfaceActor-0] Host connection pool to 127.0.0.1:9000 has completed orderly shutdown
[DEBUG] [04/05/2016 16:05:17.809] [Reproducer-akka.actor.default-dispatcher-7] [EventStream] shutting down: StandardOutLogger started
```

**Comment by ktoso** [comment]
Please provide a way to reproduce (on Akka `2.4.3`) @hochgi, or I'll have to close.
Thanks in advance!

**Comment by hochgi** [comment]
I was able to reproduce in a new project:
https://github.com/hochgi/akka-http-test
run `sbt it/it:test`
I got:
```none
[info] initiating tests
[info] HttpTest:
[info] akka-http-test
[info] - should get an empty response with http wrapper *** FAILED ***
[info]   java.lang.IllegalArgumentException: requirement failed: contentLength must be positive (use `HttpEntity.empty(contentType)` for empty entities)
[info]   at scala.Predef$.require(Predef.scala:224)
[info]   at akka.http.scaladsl.model.HttpEntity$Default.<init>(HttpEntity.scala:271)
[info]   at akka.http.scaladsl.model.HttpEntity$Strict.withSizeLimit(HttpEntity.scala:232)
[info]   at akka.http.scaladsl.model.HttpEntity$Strict.withSizeLimit(HttpEntity.scala:209)
[info]   at com.github.hochgi.util.http.SimpleHttpClient$.com$github$hochgi$util$http$SimpleHttpClient$$resToSimpleRes(SimpleHttpClient.scala:141)
[info]   at com.github.hochgi.util.http.SimpleHttpClient$$anonfun$request$1.apply(SimpleHttpClient.scala:164)
[info]   at com.github.hochgi.util.http.SimpleHttpClient$$anonfun$request$1.apply(SimpleHttpClient.scala:163)
[info]   at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
[info]   at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
[info]   at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
[info]   ...
[info] Run completed in 8 seconds, 17 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	com.github.hochgi.test.HttpTest
[error] (it/it:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 23 s, completed Apr 5, 2016 8:01:50 PM
```

**Comment by hochgi** [comment]
Oh... this fails on 2.4.2,
but appears to be working for 2.4.3 !!!
(I should have upgraded the akka version in our code - we're still working against 2.4.2)
I guess you could close the ticket, although if you weren’t aware to this bug earlier,
it might be worth the investigation.

**Comment by ktoso** [comment]
Interesting! Thanks a lot of the reproducer - those help a lot.

I'm not yet sure what could have fixed it but will want to figure it out
before closing this ticket :)

**Comment by ktoso** [comment]
Hi @hochgi, thanks a lot for the nice chat this week at Scalapeno :-)
Let's close the ticket since it's working now, thanks a lot for investigating!

**Pull request closed without merging**